jetty-openid/src/main/java/org/eclipse/jetty/security/openid/OpenIdAuthenticator.java

```diff
@@ -494,7 +491,82 @@ protected String getChallengeUri(HttpServletRequest request)
     @Override
     public boolean secureResponse(ServletRequest req, ServletResponse res, boolean mandatory, User validatedUser)
     {
-        return true;
+        return req.isSecure();
+    }
+
+    private UriRedirectInfo removeAndClearCsrfMap(HttpSession session, String csrf)
+    {
+        @SuppressWarnings("unchecked")
+        Map<String, UriRedirectInfo> csrfMap = (Map<String, UriRedirectInfo>)session.getAttribute(CSRF_MAP);
+        if (csrfMap == null)
+            return null;
+
+        UriRedirectInfo uriRedirectInfo = csrfMap.get(csrf);
+        csrfMap.clear();
+        return uriRedirectInfo;
+    }
+
+    private Map<String, UriRedirectInfo> ensureCsrfMap(HttpSession session)
+    {
+        @SuppressWarnings("unchecked")
+        Map<String, UriRedirectInfo> csrfMap = (Map<String, UriRedirectInfo>)session.getAttribute(CSRF_MAP);
+        if (csrfMap == null)
+        {
+            // Create a custom Map so we can only have a limited number of request URIs saved.
+            csrfMap = new LinkedHashMap<>()
+            {
+                private static final int MAX_SIZE = 64;
+
+                @Override
+                protected boolean removeEldestEntry(Map.Entry<String, UriRedirectInfo> eldest)
+                {
+                    return size() > MAX_SIZE;
+                }
+            };
+            session.setAttribute(CSRF_MAP, csrfMap);
+        }
+        return csrfMap;
+    }
+
+    private static class UriRedirectInfo implements Serializable
```

Comment:
Are we sure this class can be `private`? How can it be restored by the jetty-session module? How does it work with JPMS, which has stricter rules?

If you don't want to expose a `private` class, use a `Map` for the 3 values that it stores.

It is important to get this right from the beginning to avoid incompatibilities when updating the Jetty version.

Reply:
@sbordet  I don't know about JPMS, but it seems that java serialization doesn't care about public,private,protected.  Our session attribute serialization calls `ObjectOutputStream.writeObject() `for each attribute value in the session. I created a test that did substantially the same thing as this code (storing a `HashMap` as the session attribute, where some of the entries in the `HashMap` were instances of a private class) and used the `FileSessionDataStore` to persist the session:  the session was serialized and deserialized without error.